Subject: xen3 TCP/UDP checksumming
From: http://xenbits.xensource.com/xen-unstable.hg (tip 13385)
Acked-by: jbeulich@novell.com

This is only a guess, based on suggestions from Keir Fraser.

Index: head-2007-01-16/net/ipv4/netfilter/ip_nat_proto_tcp.c
===================================================================
--- head-2007-01-16.orig/net/ipv4/netfilter/ip_nat_proto_tcp.c	2007-01-17 09:33:27.000000000 +0100
+++ head-2007-01-16/net/ipv4/netfilter/ip_nat_proto_tcp.c	2007-01-17 09:51:15.000000000 +0100
@@ -129,6 +129,9 @@ tcp_manip_pkt(struct sk_buff **pskb,
 	if (hdrsize < sizeof(*hdr))
 		return 1;
 
+	if (skb_checksum_setup(*pskb))
+		return 0;
+
 	nf_proto_csum_replace4(&hdr->check, *pskb, oldip, newip, 1);
 	nf_proto_csum_replace2(&hdr->check, *pskb, oldport, newport, 0);
 	return 1;
Index: head-2007-01-16/net/ipv4/netfilter/ip_nat_proto_udp.c
===================================================================
--- head-2007-01-16.orig/net/ipv4/netfilter/ip_nat_proto_udp.c	2007-01-17 09:33:27.000000000 +0100
+++ head-2007-01-16/net/ipv4/netfilter/ip_nat_proto_udp.c	2007-01-17 09:53:19.000000000 +0100
@@ -114,6 +114,9 @@ udp_manip_pkt(struct sk_buff **pskb,
 		portptr = &hdr->dest;
 	}
 
+	if (skb_checksum_setup(*pskb))
+		return 0;
+
 	if (hdr->check || (*pskb)->ip_summed == CHECKSUM_PARTIAL) {
 		nf_proto_csum_replace4(&hdr->check, *pskb, oldip, newip, 1);
 		nf_proto_csum_replace2(&hdr->check, *pskb, *portptr, newport, 0);
Index: head-2007-01-16/net/ipv4/xfrm4_output.c
===================================================================
--- head-2007-01-16.orig/net/ipv4/xfrm4_output.c	2006-11-29 22:57:37.000000000 +0100
+++ head-2007-01-16/net/ipv4/xfrm4_output.c	2007-01-17 09:52:37.000000000 +0100
@@ -48,6 +48,10 @@ static int xfrm4_output_one(struct sk_bu
 	struct xfrm_state *x = dst->xfrm;
 	int err;
 	
+	err = skb_checksum_setup(skb);
+	if (err)
+		goto error_nolock;
+
 	if (skb->ip_summed == CHECKSUM_PARTIAL) {
 		err = skb_checksum_help(skb);
 		if (err)
